### ISSUE TITLE
Fail builds on error

### ISSUE THREAD

**Issue by felixfbecker** [semantic-release/semantic-release#569]
If no NPM token is set and semantic-release is run as `script` in Travis, the build still passes and semantic-release exits with 0.

```
1.81s$ npm run semantic-release


> iterare@1.0.0 semantic-release /home/travis/build/felixfbecker/iterare

> semantic-release


[Semantic release]: Load plugin verifyConditions from @semantic-release/npm

[Semantic release]: Load plugin verifyConditions from @semantic-release/github

[Semantic release]: Load plugin verifyConditions from @semantic-release/condition-travis

[Semantic release]: Load plugin getLastRelease from @semantic-release/npm

[Semantic release]: Load plugin analyzeCommits from @semantic-release/commit-analyzer

[Semantic release]: Load plugin generateNotes from @semantic-release/release-notes-generator

[Semantic release]: Load plugin publish from @semantic-release/npm

[Semantic release]: Load plugin publish from @semantic-release/github

[Semantic release]: Run automated release from branch master

[Semantic release]: Call plugin verify-conditions

[Semantic release]: Verify authentication for registry https://registry.npmjs.org/

[Semantic release]: ENONPMTOKEN No npm token specified.



The command "npm run semantic-release" exited with 0.
```

**Comment by pvdlg** [comment]
That's the intended behavior. "Expected" errors reported with `SemanticReleaseError` are logged and exit the process with `0`. Unexpected errors reported with `Error` are logged and exit the process with `1`.

Unexpected errors are code issues, network problems or errors in the underlying dependencies.
Expected errors cover configuration issues and various situation in which no release has to be performed.

Ideally (in a future version) those expected errors should be reported to the users in a more direct way such has opening an issue or sending a notification of some sort, in addition of being logged.

The reason for this behavior is to not fail the build when a release cannot be performed. In such case the user should be made aware but not releasing is not a reason to break a build (like failing test would).
We recommend to use `semantic-release` in the `after_success` step in Travis whether we return `0` or `1` doesn't matter in that case. But not all CI have an `after_success` step.

**Comment by felixfbecker** [comment]
Strongly disagree. Why would this not be a reason to break the build? It’s a fatal CI misconfiguration. If deploys are not happening even though I’d expect them to that’s as close as it gets to the definition of a build failure.

One of the major advantages of semantic-release v11 is that it doesn’t have to be run in after_success anymore because it doesn’t use exit codes to signal ENOCHANGE. Meaning it can be executed in script / in a Travis build step and not get silent failures.

If someone really wants to ignore failed releases they can use e.g. Travis allow_failures, but the failed release should still be visible somehow, not hidden in a lig while everything is green

**Comment by pvdlg** [comment]
This actually a much broader subject than that, which involve future user reporting, potential Github application, management of multiple branches and dist-tag with reporting/prevention of some inconstant cases.

A lot of the current implementation has historic roots (such as using `SemanticReleaseError`).

This is the exception/error situation we/we'll have to handle:
1. Coding error in a plugin or wrong output
2. Unexpected Error in a dependency
3. Exceptional case for which the user doesn't need to be alerted (like no release to be done)
4. Exceptional cases for which the user has to be alerted (i.e. the commit xyz create an incoherent state of your branching/release system or a misconfiguration)

The current implementation doesn't handle that level or granularity very well. It works for now as we don't need to distinguish 3 and 4 and just log. Once we have a better reporting system we'll have to figure out a better implementation to handle 3 and 4.

In addition of that there is the question of what should (if anything) break the build.
A case like a misconfiguration for example is really debatable. Should we fail the build or should we let the build succeed and open an issue to alert the user?
If you have a very active project and a maintainer decide to try out `semantic-release` and make a misconfiguration, it's probably better to alert via an issue or some other means rather than breaking all the builds for everyone.
If your private `npm` is down, should we report that to users via an issue or a notification or should we fail every builds, prevent PR to be merged, trigger the same alerts that are triggered when a unit test fail and basically indicate "your code is broken" while it's not?

Those type of questions are related to the bigger question of the relationship between CI and release management. Some might want the release to be part of the CI: if the release fails then the CI build fails, similarly to when a unit test fails. Some might see it as something that comes after the CI: if the release fails it doesn't mean my code is broken like a failed unit test would.

Your strong disagreement is understandable, but there is also some reasons for the current implementation, mainly to be less disruptive in order to, among other things, encourage adoption.
So the current approach is: "If we have what we need to make a release and we are sure one has to be made, we'll do it for you. If not we'll let you know (via log for now) and don't break anything in your workflow".

It's a fait point though that the user can (in some cases that depends on the CI platform) use `test` or `after_sucess` step to decide if a failed release should fail the build. Still remains the question of what qualify as a failed release.

Long story short: yes error management has to be improved, but it's complex and the solution has to encompass future `semantic-release` features.

On a side note `allow_failures` would allow failures on the whole CI job, not only on the `semantic-release` call. And `after_success` is available on Travis but not necessary on every CI solutions (especially private/corporate ones).

**Comment by felixfbecker** [comment]
@pvdlg thanks for your reply, let me respond to each point individually (tldr at the end):

> which involve future user reporting, potential Github application

Imo this isn't coupled to the GitHub application, which can always have a different reporting mechanism. This is as low-level as "if there was a configuration error, make the process return a non-zero exit code". Complex projects and companies will always want to have their release process go through their CI system, through a deploy bot or whatever, and they might not even use GitHub. The CI workflow should always be a first-class citizen in semantic-release.

> This is the exception/error situation we/we'll have to handle:
> 
> 1. Coding error in a plugin or wrong output
> 2. Unexpected Error in a dependency
> 3. Exceptional case for which the user doesn't need to be alerted (like no release to be done)
> 4. Exceptional cases for which the user has to be alerted (i.e. the commit xyz create an incoherent state of your branching/release system or a misconfiguration)

The 4 error types you mentioned can be reduced to just two: Non-failure cases (ENOCHANGE etc) and failure cases (everything else). Looks like it could be achieved atm by simply using `Error` for the ENONPMTOKEN error.

> A case like a misconfiguration for example is really debatable. Should we fail the build or should we let the build succeed and open an issue to alert the user?

What would be the benefit of opening an issue if I am not using the GitHub app? CI systems already have methods to notify about broken builds, where it's easy to configure that for example only I get notified as the one who broke the build.

> If you have a very active project and a maintainer decide to try out semantic-release and make a misconfiguration, it's probably better to alert via an issue or some other means rather than breaking all the builds for everyone.

If you use CI for deployment (aka Continuous Deployment) then your builds _are_ broken for everyone, because they don't do anymore what they are supposed to do (deploy). You could make the same argument for anything that could break the build, e.g. "should a lint failure turn the build status red or just create a GitHub issue". But in reality I see no benefit from that. Multiple commits could be pushed to try to fix the config. If the build fails, it's directly tied to the commit and easy to see what commit works / which one is wrong and who is responsible. A GitHub issue would be much harder to link to the failing builds+commits and to make sure the responsible people are notified (and others not). And most importantly: The GitHub issue reporting doesn't exist yet.

> If your private npm is down, should we report that to users via an issue or a notification or should we fail every builds, prevent PR to be merged, trigger the same alerts that are triggered when a unit test fail and basically indicate "your code is broken" while it's not?

CI is more than just unit test, it is what you as a user use it for. If you use it for unit tests, it's for unit tests. If you also use it for linting, it is also for linting (and fails the build if linting fails, unless you configure it to ignore that failure). If you use it for deployment/releases, it is also for deployment/releases, and should fail the build if that part fails (unless you configured CI to ignore that failure).

> prevent PR to be merged

That is up to you to configure. If you configure CI to only run semantic-release on master, then this will not prevent PRs to be merged.

> Those type of questions are related to the bigger question of the relationship between CI and release management. Some might want the release to be part of the CI: if the release fails then the CI build fails, similarly to when a unit test fails. Some might see it as something that comes after the CI: if the release fails it doesn't mean my code is broken like a failed unit test would.

Yes, and if semantic-release exits with a non-zero exit code, everyone can configure CI to ignore that error. They can use `allow_failures`, or make it as simple as appending `|| true` to the command to ignore the exit code. But if semantic-release always exits always with 0, it's impossible to opt-in into non-silent errors. And I believe that should be the default, because it is a huge foot gun.

> Your strong disagreement is understandable, but there is also some reasons for the current implementation, mainly to be less disruptive in order to, among other things, encourage adoption.

Imo foot guns are not good for adoption, they cause frustration.

> "If we have what we need to make a release and we are sure one has to be made, we'll do it for you. If not we'll let you know (via log for now) and don't break anything in your workflow".

If I set up my workflow to do my releases in CI, but something is wrong (token expired etc), then my workflow _is_ broken. I would not want it to pretend everything is fine.

> It's a fait point though that the user can (in some cases that depends on the CI platform) use test or after_sucess step to decide if a failed release should fail the build.

> On a side note allow_failures would allow failures on the whole CI job, not only on the semantic-release call. And after_success is available on Travis but not necessary on every CI solutions (especially private/corporate ones).

To use `allow_failures` you have to separate the release into a separate job/build stage of course, which is supported by Travis (build stages), CircleCI (workflows), and Buildkite (pipelines). If not, every CI provider can run a shell script and a shell script can ignore an exit code - if you really want the release to not impact your build status, that is what you want anyway so bugs don't affect it either.

> Still remains the question of what qualify as a failed release.

I think that is easy to answer: If the build ran, should have done a release or was unable to find out, but has not, that's a failed release. I.e.:
- ENOCHANGE is not a failure
- A config error like ENONPMTOKEN is a failure
- npm registry down is a failure
- git exec error is a failure
- Any other uncaught exception is a failure

TL;DR:

- I believe the current behaviour is a foot gun for new users and should not be the default.
- It would be easy enough to opt-in to it if desired.

**Comment by felixfbecker** [comment]
To give some graphic examples:

#### Circle:

![image](https://user-images.githubusercontent.com/10532611/34184274-a7a9c696-e4d3-11e7-96ac-3aa0231f496f.png)

#### Travis:

![image](https://user-images.githubusercontent.com/10532611/34184358-20513cc8-e4d4-11e7-8f16-81c821bdef71.png)

#### Buildkite:

![image](https://user-images.githubusercontent.com/10532611/34184435-9f7ecae2-e4d4-11e7-8f33-06ba82d3906c.png)

If the release was actually not possible e.g. because my npm token was not set or expired, as a user I would expect to see `❌  Release`, not `✅ Release`.

**Comment by pvdlg** [comment]
The big problem for now is that the travis plugin throw errors that shouldn't fail the build (like if we are on a PR or we are not running on the build leader).

Some of the changes in #563 will make that easier, so I'll tackle the error management just after that.

The objective is to have any actual error to throw `1` as requested in this issue.

**Comment by felixfbecker** [comment]
@pvdlg wouldn't it be enough to always use `SemanticReleaseError` for non-breaking errors and `Error` for breaking errors?

**Comment by pvdlg** [comment]
I'd rather not for 2 reasons:
1. In a future feature I'd like to report `SemanticReleaseError` to the user via opening issues. While `Error` will just fail the build as they do now.
2. That would involve changing all the plugins, including third-parties ones. And change them again once we want to report functional errors to the users via issues

After #563 the only thing that should be left in the `conditions-travis` should be to check that we run on the last job. We'll be able to move that somewhere else or to make a few changes to report the fact that we are not on the build leader via another mean than `SemanticReleaseError`

**Comment by felixfbecker** [comment]
Why would it require changing all the plugins? As I understand it the current behaviour is that SemanticReleaseError does _not_ fail the build, so that behaviour would be maintained, and normal Errors should never be thrown unless because of a bug.

Maybe give it a boolean property like `expected` then?

**Comment by pvdlg** [comment]
> Why would it require changing all the plugins?

Because currently all the plugins return `SemanticReleaseError` for config error, invalid tokens etc.

> Maybe give it a boolean property like expected then?

The idea would be instead to have the travis plugin to return `false` instead of throwing a `SemanticReleaseError`. Or maybe to handle that outside of `semantic-releasse` by running on Travis (for cases without Build Stages) something like `$ travis-deploy-once && semantic-release`.

But I'd rather tackle that just after #563. #563 is well on its way with changes on every plugins. Doing what you are proposing in this issue (basically all failures return `1`) would be much easier then.

**Comment by felixfbecker** [comment]
I mean, yeah, it would require updating all plugins to achieve the proposed behaviour. Otherwise this would be a breaking change.

A bit confused why you are talking about the Travis plugin, just to be clear I am talking about `@semantic-release/npm`.

> The idea would be instead to have the travis plugin to return false instead of throwing a SemanticReleaseError.

I don't think that is a good idea because `false` does not contain any information anymore what exactly happened, no message, no error code etc. Yes, the plugin could log itself, but if used programmatically (I've seen this) the caller would have no way to catch and handle specific errors manually.

**Comment by pvdlg** [comment]
Every plugin currently returns `SemanticReleaseError` when a config/authorization error happen. Basically all errors for which the user is responsible. All this errors will be reported to the user at some point in the future ("Please fix the config", "Please setup the auth", etc..).

There is one exception: [condition-travis]https://github.com/semantic-release/condition-travis) which returns `SemanticReleaseError` when we run on a PR build or on a job which is not the build leader. Those errors should not fail the build nor bother the user.

In this specific situation it's fine to return `false` because it's not an error and there is nothing more to report. That's part of it works. We can log in order to help debugging but there is nothing else to report. Returning false means modifying the way the `verifyingCondition` plugin works though.

In conclusion, this is the [condition-travis]https://github.com/semantic-release/condition-travis) plugin that prevent to modify the core to just return `1` on `SemanticReleaseError`.

As mentioned before after #563 is done, that would be much easier to solve the real root cause of the problem the [condition-travis]https://github.com/semantic-release/condition-travis) plugin.

**Comment by felixfbecker** [comment]
There is also the ENOCHANGE error, right?

**Comment by pvdlg** [comment]
> There is also the ENOCHANGE error, right?

That's part of the core so it's easier. It will be handled as part of #563

**Comment by felixfbecker** [comment]
Not sure I follow. Why is that part of the core and condition-travis not? Doesn't every third-party `analyzeCommits` plugin also use such an error?

**Comment by pvdlg** [comment]
No `analyzeCommits` was done in a "smarter" way and allow to return a falsy value when no release has to be done. this is the code that get the falsy value and throw  ENOCHANGE.

**Comment by chasingmaxwell** [comment]
This is problematic for me as well. I'd prefer my builds to fail loudly if no release is performed. Using Travis build stages, the release job isn't run unless it is fully expected to release. All conditions are verified in travis config. I use the `condition-noop` plugin to prevent the verifyConditions step from preventing a release. For this reason, I'm going to have to stick with ^8.0.0 for now. It'd be great if this behavior was configurable with a cli flag. Something like: `--fail-without-release`.

**Comment by felixfbecker** [comment]
I just had another case of this. The only reason I found out was because [a user tweeted at me whether we could get another release](https://twitter.com/jens1o/status/944646316147257344). Which made me think, weird, that should have already been auto-released - and [the build had passed](https://travis-ci.org/felixfbecker/php-language-server/builds/320497231).

Turns out the release was not working because Docker was updated in Travis:
```
$ npm run semantic-release
> php-language-server@0.0.0-development semantic-release /home/travis/build/[secure]/php-language-server
> semantic-release
[Semantic release]: Load plugin verifyConditions from @semantic-release/github
[Semantic release]: Load plugin verifyConditions from semantic-release-docker
[Semantic release]: Load plugin getLastRelease from @semantic-release/last-release-git-tag
[Semantic release]: Load plugin analyzeCommits from @semantic-release/commit-analyzer
[Semantic release]: Load plugin generateNotes from @semantic-release/release-notes-generator
[Semantic release]: Load plugin publish from @semantic-release/github
[Semantic release]: Load plugin publish from semantic-release-docker
[Semantic release]: Run automated release from branch master
[Semantic release]: Call plugin verify-conditions
[Semantic release]: Verify Github authentication
unknown shorthand flag: 'e' in -e=[secure]
See 'docker login --help'.
[Semantic release]: undefined docker login failed
The command "npm run semantic-release" exited with 0.
```
I would have never found out if I didn't get tweeted at.

And to be clear, I would not want a GitHub issue for this. I don't subscribe to all opened GitHub issues, it's just way to many. I check them every once in a while. But if my build/release is failing, I want to get an email for that and fix it asap.

**Comment by AndyOGo** [comment]
@felixfbecker 
@pvdlg 
What if error-level and related exit-codes could be configured, to match the consumers needs?
